core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -450,34 +450,41 @@ class KafkaApis(val requestChannel: RequestChannel,
       val convertedPartitionData = {
         responsePartitionData.map { case (tp, data) =>
 
-          // We only do down-conversion when:
-          // 1. The message format version configured for the topic is using magic value > 0, and
-          // 2. The message set contains message whose magic > 0
-          // This is to reduce the message format conversion as much as possible. The conversion will only occur
-          // when new message format is used for the topic and we see an old request.
-          // Please note that if the message format is changed from a higher version back to lower version this
-          // test might break because some messages in new message format can be delivered to consumers before 0.10.0.0
-          // without format down conversion.
-          val convertedData = if (versionId <= 1 && replicaManager.getMagic(tp).exists(_ > Record.MAGIC_VALUE_V0) &&
-            !data.records.hasMatchingShallowMagic(Record.MAGIC_VALUE_V0)) {
-            trace(s"Down converting message to V0 for fetch request from $clientId")
-            val downConvertedRecords = data.records.toMessageFormat(Record.MAGIC_VALUE_V0, TimestampType.NO_TIMESTAMP_TYPE)
-            FetchPartitionData(data.error, data.hw, downConvertedRecords)
-          } else data
-
-          tp -> new FetchResponse.PartitionData(convertedData.error, convertedData.hw, convertedData.records)
+          // Down-conversion of the fetched records is needed when the stored magic version is
+          // greater than that supported by the client (as indicated by the fetch request version). If the
+          // configured magic version for the topic is less than or equal to that supported by the version of the
+          // fetch request, we skip the iteration through the records in order to check the magic version since we
+          // know it must be supported. However, if the magic version is changed from a higher version back to a
+          // lower version, this check will no longer be valid and we will fail to down-convert the messages
+          // which were written in the new format prior to the version downgrade.
+          val convertedData = replicaManager.getMagic(tp) match {
+            case Some(magic) if magic > 0 && versionId <= 1 && !data.records.hasCompatibleMagic(RecordBatch.MAGIC_VALUE_V0) =>
+              trace(s"Down converting message to V0 for fetch request from $clientId")
+              FetchPartitionData(data.error, data.hw, data.records.downConvert(RecordBatch.MAGIC_VALUE_V0))
+
+            case Some(magic) if magic > 1 && versionId <= 3 && !data.records.hasCompatibleMagic(RecordBatch.MAGIC_VALUE_V1) =>
+              trace(s"Down converting message to V1 for fetch request from $clientId")
+              FetchPartitionData(data.error, data.hw, data.records.downConvert(RecordBatch.MAGIC_VALUE_V1))
+
+            case _ => data
+          }
```

Comment:
Yes, I agree. There are probably ways we can do it so that we keep so visibility into the fact that the down-conversion is happening behind the scenes (maybe just comments).